lta/transfer/globus.py

```diff
@@ -1,108 +1,196 @@
+
+    async def transfer_file(
+        self,
+        *,
+        source_path: Path,
+        dest_path: Path,
+    ) -> uuid.UUID | str:
+        """
+        Transfer a single file via Globus Transfer and block until completion.
+
+        :param source_path: Absolute path on the source collection.
+        :param dest_path: The filepath on the destination collection.
+
+        :returns: Globus task_id for the submitted transfer.
+        """
+        if not os.path.isabs(source_path):
+            raise ValueError(f"source_path must be absolute: {source_path}")
+
+        # do transfer
+        tdata = self.make_transfer_document(source_path, dest_path)
+        task_id = await self._submit_transfer(tdata)
+
+        return task_id
+
+    async def wait_for_transfer_to_finish(
+            self,
+            task_id: uuid.UUID | str,
+    ) -> dict[str, Any]:
+        """Wait (forever) for transfer to finish.
+
+        Return:
+            The successful task object.
+
+        Raises:
+            'GlobusTransferFailedException' if the transfer failed (FAILED or INACTIVE).
+
+        NOTE: 'globus_sdk.TransferClient.task_wait()' is *NOT* async, so we must diy
+        """
+        first = True
+        while True:
+
+            # sleep each time after first iteration
+            if first:
+                first = False
             else:
-                cmd = ['grid-proxy-init']
-            cmd.extend(['-debug', '-pwstdin', '-valid', f'{int(self.cfg["GLOBUS_PROXY_DURATION"])+1}:0'])
-            if 'GLOBUS_PROXY_OUTPUT' in self.cfg and self.cfg['GLOBUS_PROXY_OUTPUT']:
-                cmd.extend(['-out', cast(str, self.cfg['GLOBUS_PROXY_OUTPUT'])])
-            inputbytes = (cast(str, self.cfg['GLOBUS_PROXY_PASSPHRASE']) + '\n').encode('utf-8')
-            p = subprocess.run(cmd, input=inputbytes, capture_output=True, timeout=60, check=False)
-            logger.info('proxy cmd: %r', p.args)
-            logger.info('stdout: %s', p.stdout)
-            logger.info('stderr: %s', p.stderr)
-            if 'GLOBUS_PROXY_VOMS_VO' in self.cfg and self.cfg['GLOBUS_PROXY_VOMS_VO']:
-                for line in p.stdout.decode('utf-8').split('\n'):
-                    if line.startswith('Creating proxy') and line.endswith('Done'):
-                        break  # this is a good proxy
-                else:
-                    raise Exception('voms-proxy-init failed')
-            elif p.returncode > 0:
-                raise Exception('grid-proxy-init failed')
-
-    def get_proxy(self) -> Any:
-        """Get the proxy location."""
-        if 'GLOBUS_PROXY_OUTPUT' in self.cfg and self.cfg['GLOBUS_PROXY_OUTPUT']:
-            return self.cfg['GLOBUS_PROXY_OUTPUT']
-        FNULL = open(os.devnull, 'w')
-        return subprocess.check_output(['grid-proxy-info', '-path'],
-                                       stderr=FNULL).decode('utf-8').strip()
+                await asyncio.sleep(self._env.GLOBUS_POLL_INTERVAL_SECONDS)
```

Comment:
Maybe remove the `first` logic? I can see the time savings, but it's on the order of a poll interval; my instinct is to sleep straight away and eat any time cost, with the benefit of simpler code / control flow.

This is just my opinion; not an error, not something you have to change.
If you agree, change it. If not, keep it as-is.

Reply:
For LTA's actual use cases, I agree with you that this is a pre-optimization. However, it is helpful for the `examples/globus_transfer_file.py` script since it can transfer files of any size.